airsonic-main/src/main/java/org/airsonic/player/service/MediaScannerService.java

```diff
@@ -145,68 +173,76 @@ public synchronized void scanLibrary() {
-                doScanLibrary();
-                playlistService.importPlaylists();
-                mediaFileDao.checkpoint();
-            }
-        };
-
-        thread.setPriority(Thread.MIN_PRIORITY);
-        thread.start();
+        ForkJoinPool pool = new ForkJoinPool(scannerParallelism, mediaScannerThreadFactory, null, true);
+        
+        CompletableFuture.runAsync(() -> doScanLibrary(), pool)
+                .thenRunAsync(() -> playlistService.importPlaylists(), pool)
+                .thenRunAsync(() -> mediaFileDao.checkpoint(), pool)
+                .thenRun(() -> pool.shutdown())
+                .thenRun(() -> scanning = false);
     }
 
     private void doScanLibrary() {
         LOG.info("Starting to scan media library.");
-        Instant lastScanned = Instant.now().truncatedTo(ChronoUnit.SECONDS);
+        Instant lastScanned = Instant.now();
         LOG.debug("New last scan date is {}", lastScanned);
 
         try {
-
             // Maps from artist name to album count.
-            Map<String, Integer> albumCount = new HashMap<String, Integer>();
+            Map<String, AtomicInteger> albumCount = new ConcurrentHashMap<>();
+            Map<String, Artist> artists = new ConcurrentHashMap<>();
+            Map<MediaFile, Album> albums = new ConcurrentHashMap<>();
             Genres genres = new Genres();
 
-            scanCount = 0;
+            scanCount.set(0);
             statistics.reset();
 
             mediaFileService.setMemoryCacheEnabled(false);
             indexManager.startIndexing();
 
-            mediaFileService.clearMemoryCache();
-
             // Recurse through all files on disk.
-            for (MusicFolder musicFolder : settingsService.getAllMusicFolders()) {
-                MediaFile root = mediaFileService.getMediaFile(musicFolder.getPath(), false);
-                scanFile(root, musicFolder, lastScanned, albumCount, genres, false);
-            }
-
+            ForkJoinTask.invokeAll(
+                    settingsService.getAllMusicFolders()
+                        .parallelStream()
+                        .map(musicFolder -> ForkJoinTask.adapt(() -> scanFile(mediaFileService.getMediaFile(musicFolder.getPath(), false), musicFolder, lastScanned, albumCount, artists, albums, genres, false)))
+                        .collect(Collectors.toList()));
+            
             // Scan podcast folder.
             File podcastFolder = new File(settingsService.getPodcastFolder());
             if (podcastFolder.exists()) {
-                scanFile(mediaFileService.getMediaFile(podcastFolder), new MusicFolder(podcastFolder, null, true, null),
-                         lastScanned, albumCount, genres, true);
+                ForkJoinTask.invokeAll(ForkJoinTask.adapt(() -> scanFile(mediaFileService.getMediaFile(podcastFolder), new MusicFolder(podcastFolder, null, true, null),
+                         lastScanned, albumCount, artists, albums, genres, true)));
```

Comment:
This is a single task, and we _could_ just run it here by itself, this thread itself will work on
But just submit to the pool